flow/damage_context.cc

```diff
@@ -0,0 +1,139 @@
+#include "flutter/flow/damage_context.h"
+#include "flutter/flow/layers/layer.h"
+#include "third_party/skia/include/core/SkImageFilter.h"
+
+namespace flutter {
+
+std::size_t DamageContext::LayerEntry::Hash::operator()(
+    const LayerEntry& e) const noexcept {
+  size_t res = e.paint_bounds.left();
+  res = 37 * res + e.paint_bounds.top();
+  res = 37 * res + e.paint_bounds.width();
+  res = 37 * res + e.paint_bounds.height();
+  res = 37 * res + size_t(reinterpret_cast<std::uintptr_t>(e.comparator));
+  return res;
+}
+
+bool DamageContext::LayerEntry::operator==(const LayerEntry& e) const {
+  return comparator == e.comparator && paint_bounds == e.paint_bounds &&
+         std::equal(
+             mutators.begin(), mutators.end(), e.mutators.begin(),
+             e.mutators.end(),
+             [](const std::shared_ptr<Mutator>& m1,
+                const std::shared_ptr<Mutator>& m2) { return *m1 == *m2; }) &&
+         (layer.get() == e.layer.get() ||
+          comparator(layer.get(), e.layer.get()));
+}
+
+void DamageContext::InitFrame(const SkISize& tree_size,
+                              const FrameDescription* previous_frame) {
+  current_layer_tree_size_ = tree_size;
+  previous_frame_ = previous_frame;
+}
+
+void DamageContext::PushLayerEntry(const Layer* layer,
+                                   LayerComparator comparator,
+                                   const SkMatrix& matrix,
+                                   const PrerollContext& preroll_context,
+                                   size_t index) {
+  if (current_layer_tree_size_.isEmpty() || layer->paint_bounds().isEmpty()) {
+    return;
+  }
+
+  LayerEntry e;
+  e.layer = layer->shared_from_this();
+  e.comparator = comparator;
+
+  SkRect bounds = layer->paint_bounds();
+  bounds.intersect(preroll_context.cull_rect);
+  e.paint_bounds = matrix.mapRect(bounds);
+
+  for (auto i = preroll_context.mutators_stack.Begin();
+       i != preroll_context.mutators_stack.End(); ++i) {
+    auto type = (*i)->GetType();
```

Comment:
A performance idea. Rather than keep the mutators in a stack, keep them in a manually linked list starting from most recent. This loop then becomes `e.mutators = preroll_context.mutators`. No copy is needed because each chain will be unique even if future nodes mutate `preroll_context.mutators` and once a linked-list-mutator is added to the chain, its next pointer never changes. It can be popped, and another mutator can be added in its place, but it will never point at a different "next mutator".

Reply:
(with shared pointers in the linked list of course)